File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        if (str1 + str2 != str2 + str1) { return ""; }
    

        int gcdLength = gcd(str1.length(), str2.length());

        return str1.substr(0, gcdLength);
    }

private:
    int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
};

File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.rs
impl Solution {
    pub fn gcd_of_strings(str1: String, str2: String) -> String {
        if str1.clone() + &str2 != str2.clone() + &str1 {
            return "".to_string();
        }

        let gcd_length = Self::gcd(str1.len(), str2.len());
        str1[0..gcd_length].to_string()
    }

    fn gcd(a: usize, b:usize) -> usize {
        let mut a = a;
        let mut b = b;

        while b != 0 {
            let temp = b;
            b = a % b;
            a = temp;
        }

        a
    }
}